src/commands/src/helpers/index.ts

```diff
@@ -0,0 +1,279 @@
+      // accidentally resolve if the second command somehow still
+      // exits without error
+      if (hasRejected) return;
+
+      if (exitCode !== 0) {
+        console.error(`ERROR! The command ${cmdToParams.command} closed with an exit code other than 0: ${exitCode}.`);
+        console.error('Arguments provided: ', cmdToParams.args);
+        console.error('Options provided: ', cmdToParams.options);
+
+        if (outputText) {
+          console.error('output', outputText.join());
+        }
+
+        if (errorText) {
+          console.error('error', errorText.join());
+        }
+
+        reject(new ExecuteCommandError('The command failed', { exitCode, outputText, errorText }));
+        return;
+      }
+
+      resolve({
+        outputText: outputText.join(),
+        errorText: errorText.join(),
+      });
+    });
+  });
+}
+
+/**
+ * A promisified way to execute CLI commands. This approach uses spawn instead of exec, which
+ * is a safer way of executing CLI commands. Also, spawn allows us to stream input and output in real-time.
+ * @param {string} command The CLI command we want to execute
+ * @param {string[] | undefined} args Arguments we want to provide to the command
+ * @param {SpawnOptions | undefined} options Options to configure the spawn function
+ * @param {fs.WriteStream | undefined} writeStream A writable stream object to pipe output to.
+ * For example, we can `mimic ls >> directory.txt` by creating a `WriteStream` object to write to
+ * `directory.txt`, and then provide the `WriteStream` so that we can pipe the output from the `ls`
+ * command to the `WriteStream`.
+ * @returns {Promise<CliCommandResponse>} An object containing the CLI output from `stdout` and `stderr`.
+ * stdout is the `outputText` property, and `stderr` is the `errorText` property.
+ */
+export async function executeCliCommand({
+  command,
+  args = [],
+  options = {},
+  writeStream,
+}: CliCommandParams): Promise<CliCommandResponse> {
+  return new Promise((resolve, reject) => {
+    const outputText: string[] = [];
+    const errorText: string[] = [];
+
+    const executedCommand = spawn(command, args, options);
+
+    if (writeStream) executedCommand.stdout?.pipe(writeStream);
+    executedCommand.stdout?.on('data', (data: Buffer) => {
+      outputText.push(data.toString());
+    });
+
+    executedCommand.stderr?.on('data', (data: Buffer) => {
+      errorText.push(data.toString());
+    });
+
+    executedCommand.on('error', (err) => {
+      reject(new ExecuteCommandError('The command failed', err));
+    });
+
+    executedCommand.on('close', (exitCode) => {
+      if (writeStream) writeStream.end();
+
+      if (exitCode !== 0) {
+        console.error(`ERROR! The command ${command} closed with an exit code other than 0: ${exitCode}.`);
+        console.error('Arguments provided: ', args);
+        console.error('Options provided: ', options);
+
+        if (outputText) {
+          console.error(outputText.join());
```

Comment:
@branberry just out of curiosity, is there a reason why if there is `outputText` we are passing that to a `console.error`? Is there a case where this would provide an error message? I thought `errorText` was for capturing error messages.

Reply:
This is because a command can error out, but it still might have outputted to `stdout`. This is so we are able to see everything that was logged by this command. For example, if a command fails halfway through, there will be non-error logs that could be helpful to see, so I output it in case it does exist. I probably should check for `outputText.length` now that I think about it though.

Reply:
Thanks for the explanation. I agree that it would probably make more sense to check the length because an empty array will still be truthy. Would it also make sense to maybe use `console.log` or `console.info` since `console.error` might be misleading because that output wasn't an error rather it was more information about at what point the command had an issue?

Reply:
Yeah! I think that makes sense. I'll change this to be a `console.log` instead.